src/Bedrock.Framework.Experimental/Protocols/Kafka/Messaging/KafkaMessageDescriptor.cs

```diff
@@ -0,0 +1,109 @@
+using System.Diagnostics;
+using System.Linq;
+using System;
+using System.Buffers;
+using Bedrock.Framework.Protocols.Kafka.Messaging.SchemaTypes;
+
+namespace Bedrock.Framework.Protocols.Kafka.Messaging
+{
+    /// <summary>
+    /// Defines an API message that can be used as part of the wire protocol between a Kafka client and broker.
+    /// </summary>
+    /// <typeparam name="TRequestSchema">The request schema type.</typeparam>
+    /// <typeparam name="TResponseSchema">The response schema type.</typeparam>
+    /// <remarks>
+    /// A message descriptor brings together a request and response schema type for a version of a Kafka API
+    /// message that is implemented by a Kafka broker. The request and response schema types are inclusive of all
+    /// prior versions of the message up to and including the highest version supported by this protocol implementation.
+    /// The reader and writer implementations provided by a given <see cref="KafkaMessageDescriptor{TRequestSchema, TResponseSchema}"/>
+    /// will ensure forward and backward compatibility by omitting fields or supplying default values for fields
+    /// where approprirate.
+    /// 
+    /// The highest version of a message supported by both Kafka client and broker should be used.
+    /// </remarks>
+    public abstract class KafkaMessageDescriptor<TRequestSchema, TResponseSchema> : IKafkaMessageDescriptor
+    {
+        /// <summary>
+        /// Initializes the base <see cref="KafkaMessageDescriptor{TRequestSchema, TResponseSchema}"/>.
+        /// </summary>
+        protected KafkaMessageDescriptor()
+        {
+            var attribute = (KakfaMessageAttribute)GetType()
+                .GetCustomAttributes(typeof(KakfaMessageAttribute), true)
+                .FirstOrDefault();
+
+            Debug.Assert(attribute is object, $"Expected type '{GetType().Name}' to have been applied with a '{nameof(KakfaMessageAttribute)}'.");
+
+            ApiKey = attribute.ApiKey;
+            ApiVersion = attribute.Version;
+        }
+
+        /// <inheritdoc/>
+        public ApiKey ApiKey { get; private set; }
+
+        /// <inheritdoc/>
+        public short ApiVersion { get; private set; }
+
+        /// <inheritdoc/>
+        public IKafkaMessageWriter GetRequestWriter()
+        {
+            return new KafkaMessageWriter<TRequestSchema>(WriteRequest);
+        }
+
+        /// <inheritdoc/>
+        public IKafkaMessageReader GetRequestReader()
```

Comment:
Does it need these two? If this is just for a client then it never needs to read a request or write a response?
Unless you plan on writing a server!?! 😄

Reply:
If that is true... then you could directly implement IKafkaMessageWriter/Reader and remove the need for the nested types and the extra objects?

Reply:
Yes, I think _future_ work could include a server element. Kafka proxy? In Memory server for testing? Umm, I'm sure we could think or some more. 

> ... then you could directly implement IKafkaMessageWriter/Reader and remove the need for the nested types and the extra objects?

Implementing the interfaces directly for both request/response types doesn't distinguish the methods enough to be overloaded. Hence I've had to call out the creation of a Reader/Writer for the request/response.

Reply:
yeah... was just looking for ways to avoid creating many objects per request...

As we only need to write/read one at a time. Maybe these nested types could be singletons on the connection, then substitute the delegates for each req/res ?

Again, just premature optimization. It'll come out in the wash later 😄

Reply:
Could possibly need some massaging!